ntoskrnl/kdbg/kdb_cli.c

```diff
@@ -1052,16 +1052,47 @@ KdbpTrapFrameFromPrevTss(
 }
 
 VOID __cdecl KiTrap02(VOID);
-VOID FASTCALL KiTrap03Handler(IN PKTRAP_FRAME);
+VOID __cdecl KiTrap03(VOID);
 VOID __cdecl KiTrap08(VOID);
 VOID __cdecl KiTrap09(VOID);
+VOID __cdecl KiTrap02Handler(VOID);
+VOID FASTCALL KiTrap03Handler(IN PKTRAP_FRAME);
+VOID __cdecl KiTrap08Handler(VOID);
+VOID FASTCALL KiTrap09Handler(IN PKTRAP_FRAME TrapFrame);
 
+//
+// FIXME: Should better be renamed: "DoWeHaveAParentTSS"
+//
 static BOOLEAN
 KdbpInNmiOrDoubleFaultHandler(
     ULONG_PTR Address)
 {
-    return (Address > (ULONG_PTR)KiTrap02 && Address < (ULONG_PTR)KiTrap03Handler) ||
-           (Address > (ULONG_PTR)KiTrap08 && Address < (ULONG_PTR)KiTrap09);
+#if 0 /* This fixed code works too, but it's a HACK */
+    // FIXME BUG! Nothing ensures that the C compiler correctly orders the C trap handlers!!
+    // Bug fix PR #1997
+    return (Address >= (ULONG_PTR)KiTrap02 && Address < (ULONG_PTR)KiTrap03) ||
+           (Address >= (ULONG_PTR)KiTrap02Handler && Address < (ULONG_PTR)KiTrap03Handler) ||
+           (Address >= (ULONG_PTR)KiTrap08 && Address < (ULONG_PTR)KiTrap09) ||
+           (Address >= (ULONG_PTR)KiTrap08Handler && Address < (ULONG_PTR)KiTrap09Handler);
+#else
+    USHORT Sel;
+    Sel = Ke386GetTr();
+
+    /* Selector validity check */
+    if ((Sel & (sizeof(KGDTENTRY) - 1)) ||
+        (Sel < sizeof(KGDTENTRY)))
+    {
+        return FALSE;
+    }
+
+    /* Check whether this is the base boot TSS, if so we must not have any parent TSS */
+    if (Sel == KGDT_TSS || Sel == 0x00)
+        return FALSE;
+
+    /* We are in another TSS, return TRUE so as to call KdbpTrapFrameFromPrevTss() to retrieve the previous TSS */
+    // NOTE: Sel == KGDT_DF_TSS or KGDT_NMI_TSS for Double-Fault or NMI TSS.
+    return TRUE;
```

Comment:
It appears this technique seems to work. With extra added DPRINTing, a forced double-page fault (which trigger a Task Switch) shows the following:
```
(ntoskrnl/mm/ARM3/mminit.c:1452) HAL I/O Mapping at FFFE0000 is unsafe

*** Fatal System Error: 0x0000007f
                       (0x00000008,0x8009C000,0x00000000,0x00000000)


Entered debugger on embedded INT3 at 0x0008:0x80953368.
kdb:> bt
Eip:
<80953369>
Frames:
<808899b0>
KdbpInNmiOrDoubleFaultHandler(0x808899B0) - Sel = 0x50
    KiTrap02(0x809993CC) - KiTrap03(0x80999548)
    KiTrap02Handler(0x809993D4) - KiTrap03Handler(0x809995D8)
    KiTrap08(0x8099A140) - KiTrap09(0x8099A1D8)
    KiTrap08Handler(0x8099A148) - KiTrap09Handler(0x8099A268)
TRUE!
Couldn't access memory at 0x809F9FCC!
kdb:>
```

Reply:
Better example:
```
*** Fatal System Error: 0x0000007f
                       (0x00000008,0x8009C000,0x00000000,0x00000000)


Entered debugger on embedded INT3 at 0x0008:0x80953588.
kdb:> bt
[Active TSS 0x0050 @ 0x80A10CA0]
Eip:
<ntoskrnl.exe:153589 (sdk/lib/rtl/i386/debug_asm.S:57 (RtlpBreakWithStatusInstruction))>
Frames:
<ntoskrnl.exe:899ab (ntoskrnl/ke/bug.c:1136 (KeBugCheckWithTf))>
<ntoskrnl.exe:19a1d3 (ntoskrnl/ke/i386/traphdlr.c:917 (KiTrap08Handler))>
<ntoskrnl.exe:19a140 (:0 (KiTrap08))>
[Parent TSS 0x0028 @ 0x8009C000]
<ntoskrnl.exe:19c5b3 (:0 (KiTrapReturn))>
```